clients/src/test/java/org/apache/kafka/clients/MockClient.java

```diff
@@ -336,6 +336,12 @@ public List<ClientResponse> poll(long timeoutMs, long now) {
             copy.add(response);
         }
 
+        if (copy.isEmpty()) {
+            // Simulate time advancing. If no responses are received, then we know that
+            // we waited for the whole timeoutMs.
+            time.sleep(timeoutMs);
```

Comment:
I noticed when developing my test that repeated calls to `poll()` with `timeoutMs == 10` never resulted in `time.milliseconds()` advancing, so the fictitious `Node` in my test never became ready (time never advanced past the throttled time). That's why I made this change; however, it broke all sorts of things.

Reply:
I guess most test don't expect that time advanced "automatically"... It think we need to remove this.

I think there two possibilities: start a background thread in your test before you call `runOnce()` and let the background thread advance mockTime (maybe too complex?), or, make the `MockTime` object more advanced and add some "auto-time-advance" feature, ie, each time `milliseconds()` is called, advance time by 1ms (or some other value passed into MockTime -- only your test would enable this feature.

Reply:
Thanks Matthias. I did something somewhat inspired by your comment. Ready for review.

Reply:
Should work, too, but why don't you just create `new MockTime(1L)` and use the already existing "auto-tick" feature?

Reply:
@mjsax the problem is that the following test still passes on `trunk` using `new MockTime(1L)`. However, the test that I have in the PR as it stands passes on my branch but fails on trunk:
```
    @Test
    public void senderThreadShouldNotGetStuckWhenThrottledAndAddingPartitionsToTxn() {
        // We want MockClient#poll() to advance time so that eventually the backoff expires.
        // client.advanceTimeDuringPoll(true);
        time = new MockTime(1L);

        ProducerIdAndEpoch producerIdAndEpoch = new ProducerIdAndEpoch(123456L, (short) 0);
        apiVersions.update("0", NodeApiVersions.create(ApiKeys.INIT_PRODUCER_ID.id, (short) 0, (short) 3));
        TransactionManager txnManager = new TransactionManager(logContext, "testUnresolvedSeq", 60000, 100, apiVersions);

        setupWithTransactionState(txnManager);
        doInitTransactions(txnManager, producerIdAndEpoch);

        int throttleTimeMs = 1000;
        long startTime = time.milliseconds();
        Node nodeToThrottle = metadata.fetch().nodeById(0);
        client.throttle(nodeToThrottle, throttleTimeMs);

        // Verify node is throttled a little bit. In real-life Apache Kafka, we observe that this can happen
        // as done above by throttling or with a disconnect / backoff.
        long currentPollDelay = client.pollDelayMs(nodeToThrottle, startTime);
        assertTrue(currentPollDelay > 0);
        assertTrue(currentPollDelay <= throttleTimeMs);

        txnManager.beginTransaction();
        txnManager.maybeAddPartition(tp0);

        assertFalse(txnManager.hasInFlightRequest());
        sender.runOnce();
        assertTrue(txnManager.hasInFlightRequest());

        long totalTimeToRunOnce = time.milliseconds() - startTime;

        // It should have blocked roughly only the backoffTimeMs and some change.
        assertTrue(totalTimeToRunOnce < REQUEST_TIMEOUT);
    }
```